src/dnf-plugins/product-id/test-product-id.c

```diff
@@ -93,11 +94,21 @@ typedef struct {
 } handleFixture;
 
 void setup(handleFixture *fixture, gconstpointer testData) {
+    GError *error = NULL;
     (void)testData;
     fixture->dnfContext = dnf_context_new();
+    dnf_context_setup(fixture->dnfContext, NULL, &error);
     PluginMode mode = PLUGIN_MODE_CONTEXT;
-    // This is never explicitly called
-    fixture->handle = pluginInitHandle(1, mode, (void*)fixture->dnfContext);
+    // When plugin product-id is called by libdnf application, then pluginInitHandle is called,
+    // but this function pluginInitHandle should not be never called by other application
+    // directly, because only libdnf can set DNF context properly. We have to set attributes
+    // of handle using following way.
+    // fixture->handle = pluginInitHandle(1, mode, (void*)fixture->dnfContext);
+    fixture->handle = g_malloc(sizeof(PluginHandle));
+    if(fixture->handle != NULL) {
+        fixture->handle->version = 1;
+        fixture->handle->mode = mode;
+    }
```

Comment:
`g_malloc` invokes `abort()` on memory failures [1], so this is effectively dead code.

[1] https://developer.gnome.org/glib/stable/glib-Memory-Allocation.html

Reply:
Sorry, I meant that the check itself is not useful, as it will be always true.

Reply:
Sure, `g_malloc()` invokes `abort()` on memory failures. Can `g_malloc()` return `NULL`? Yes, it can.

Reply:
> Sure, `g_malloc()` invokes `abort()` on memory failures. Can `g_malloc()` return `NULL`? Yes, it can.

No, it cannot. Please read the linked documentation:

> If any call to allocate memory using functions g_new(), g_new0(), g_renew(), g_malloc(), g_malloc0(), g_malloc0_n(), g_realloc(), and g_realloc_n() fails, the application is terminated. **This also means that there is no need to check if the call succeeded.** On the other hand, the g_try_...() family of functions returns NULL on failure that can be used as a check for unsuccessful memory allocation. The application is not terminated in this case.

(emphasis mine)

Reply:
I do read documentation. It can returns `NULL`, when argument is 0 (`sizeof()` returns 0). I think it is good to have extra check than buggy code and we speak about code for unit testing.

Reply:
For `sizeof(PluginHandle)` to be 0 then the `PluginHandle` struct must be empty, which is undefined behaviour (or even invalid standard) in C (it is allowed in C++, though). So no, this is not a concern.

Reply:
```c
#include <glib.h>
#include <stdio.h>
#include <stdlib.h>


typedef struct Foo {
} Foo;


int main(void)
{
	struct Foo *foo;
	printf("Foo size: %ld\n", sizeof(Foo));
	foo = g_malloc(sizeof(Foo));
	printf("foo: %p\n", (void*)foo);
	if(foo != NULL) {
		g_free(foo);
	}
	return EXIT_SUCCESS;
}

```

```
gcc -o foo foo.c `pkgconf --cflags --libs glib-2.0` -Wall -Wextra -Wno-long-long -std=c99
```

```
./foo
Foo size: 0
foo: (nil)
```

Is `PluginHandle` defined in our code? No, it isn't. Could it be anything? Yes, it can.

Reply:
> ```c
> typedef struct Foo {
> } Foo;
> ```

This is undefined behaviour in C:
https://stackoverflow.com/questions/53952097/empty-structs-in-c

> Is PluginHandle defined in our code? No, it isn't. Could it be anything? Yes, it can.

If they want to make sure their code works, then they better not do this (also logically there is no reason for this).